bin/git-team

```diff
@@ -0,0 +1,116 @@
+# Script to use to save space on target by "not" cloning "shallow" repo with full history and depth=1
+# Parameters:
+#   - `-o` git operation to process - default to `pull`
+#       - For operation other than `pull` shallow repository are deleted and re-cloned to target `branch`
+#       - op can be 'clone', 'checkout', or 'pull';
+#         NB: for 'pull', the branches parameter is not used.
+#   - `-c` repository category to process - team|shallow - default to `shallow`
+#   - `-b` branch to apply operation onto - no effect on `pull` operation - default to `main`
+#   - `-f` fall back branch if `branch` doesn't exists - no effect on `pull` operation - default to `main`
+#   - `-s` shallow repositories to process
+#   - `-p` production repositories to process
+#   - `-p|-s` can be either of:
+#       - "Ensembl" xxx repo name - ex: ensembl-production
+#       - "Git Ensembl config" as define in `./git-ensembl` command tool - ex: production
+#       - "File path" containing a list of repositories to process - one per line
+
+function usage {
+  echo "Usage: $0 [-o git_operation] [-c shallow|team] [-b branch] [-f default_branch] [-t target_dir] [-s shallow_file|repo|config] [-p team_file|repo|team_config]"
+  echo "Example: $0 -o clone -c shallow -s ~/shallow.repo"
+  exit 1
+}
+shallow=()
+prod=()
+while getopts 'o:c:s:p:t:b:f:d' opt; do
+  case ${opt} in
+    o) op=${OPTARG} ;;
+    b) branch=${OPTARG} ;;
+    d) default_branch=${OPTARG} ;;
+    c) category=${OPTARG} ;;
+    t) target_dir=${OPTARG} ;;
+    s) shallow=${OPTARG} ;;
+    p) prod=${OPTARG} ;;
+    *) usage ;;
+  esac
+done
+# repo configuration with file
+if [[ -f "$shallow" ]]; then
+  shallow=(`cat $shallow`)
+fi
+if [[ -f "$prod" ]]; then
+  prod=(`cat $prod`)
+fi
+
+# Ensembl shallow repositories (depth=1)
+branch=${branch:-"main"}
+default_branch=${default_branch:-"main"}
+target_dir=${target_dir:-"${PWD}"}
+
+if [[ ! -d $target_dir ]]; then
+  mkdir -p $target_dir
+fi
+
+category=${category:-'shallow'}
+op=${op:-'pull'}
+ignore=( "${shallow[@]/#/--ignore_module }" )
+dir=$(pwd)
+cd ${target_dir}
+echo "Processing '${op}' (cat:${category}) from ${PWD} ..."
+echo "=================================================="
+# Using git-tools for pull will update all tags/branches,
+# which we don't want for our shallow copies.
+if [[ $op == 'pull' ]]; then
+  if [[ $category == 'shallow' ]]; then
+    for repo in "${shallow[@]}"; do
+      echo "git -C ${target_dir}/${repo} pull"
+      echo "-----------------------------------------"
+      git -C ${target_dir}/${repo} pull
+    done
+  else
+    # shellcheck disable=SC2068
+    echo "git ensembl --$op ${ignore[@]} ${prod[@]}"
+    echo "-----------------------------------------"
+    git ensembl --$op ${ignore[@]} ${prod[@]}
+  fi
+else
+  # Cleanest way to get a shallow copy with git-tools is to
+  # wipe out and start from scratch.
+  if [[ $category == 'shallow' ]]; then
+    for repo in ${shallow[@]}; do
+      rm -rf "${repo}"
```

Comment:
Wonder if we should warn them that we are wiping out. What if there are any local changes? Depends on how friendly you want the script to be :-)

Reply:
The shallow category is supposed not to be updated outside or the team managing this repository, if someone ever want to propose changes following an update. He will certainly not run this command at this point. The point of this is mainly to initiate a repositories dir, not to maintain it overtime.

Reply:
Basically what we do in production (with an internal version of this script) is :
- use the `clone` operation when setting up an environment for a release.
- use the `pull` operation in our crons to maintains all repo in sync with branches.